demo/code_lense.html

```diff
@@ -0,0 +1,84 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+  <meta charset="UTF-8">
+  <title>ACE Autocompletion demo</title>
+  <style type="text/css" media="screen">
+    body {
+        overflow: hidden;
+    }
+    
+    #editor { 
+        margin: 0;
+        position: absolute;
+        top: 0;
+        bottom: 0;
+        left: 0;
+        right: 0;
+    }
+  </style>
+</head>
+<body>
+
+<pre id="editor"></pre>
+
+<script src="kitchen-sink/require.js"></script>
+<script>
+// setup paths
+require.config({paths: { "ace" : "../lib/ace"}});
+// load ace and extensions
+require(["ace/ace", "ace/ext/code_lense"], function(ace, codeLense) {
+    var editor = ace.edit("editor");
+    editor.session.setMode("ace/mode/html");
+
+    var commandId = "describeCodeLense"
+    editor.commands.addCommand({
+        name: commandId,
+        exec: function(editor, args) {
+            // services available in `ctx`
+            alert('CodeLense command called with arguments ' + args);
+        }
+    });
+
+    codeLense.setCodeLenseProvider(editor, {
+        provideCodeLenses: function(session) {
```

Comment:
How would I switch sessions? Shouldn't we attach the code lenses to the session state?

Reply:
I've added `enableCodeLense` option to the editor, if the option is enabled and new session with codeLenseProvider, (or with a mode with codeLenseProvider) is added to the editor, it will render codelenses automatically. The api is still not quite right, but i am not sure how to change it, maybe we can improve it later when we are using it.